unpick/src/main/java/net/neoforged/jst/unpick/UnpickCollection.java

```diff
@@ -0,0 +1,317 @@
+                    case Literal.Float(var f) -> f;
+                    case Literal.Double(var d) -> d;
+                    case Literal.String(var s) -> s;
+                };
+            } else if (expression instanceof ParenExpression parenExpression) {
+                return resolveConstant(parenExpression.expression, facade, scope);
+            } else if (expression instanceof CastExpression castExpression) {
+                return cast(resolveConstant(castExpression.operand, facade, scope), castExpression.castType);
+            } else if (expression instanceof UnaryExpression unaryExpression) {
+                var value = (Number) resolveConstant(unaryExpression.operand, facade, scope);
+                return switch (unaryExpression.operator) {
+                    case NEGATE -> NumberType.TYPES.get(value.getClass()).negate(value);
+                    case BIT_NOT -> value instanceof Long ? ~value.longValue() : ~value.intValue();
+                };
+            } else if (expression instanceof BinaryExpression binaryExpression) {
+                var lhs = resolveConstant(binaryExpression.lhs, facade, scope);
+                var rhs = resolveConstant(binaryExpression.rhs, facade, scope);
+
+                if (lhs instanceof Number l && rhs instanceof Number r) {
+                    var type = NumberType.TYPES.get(l.getClass());
+                    return switch (binaryExpression.operator) {
+                        case ADD -> type.add(l, r);
+                        case DIVIDE -> type.divide(l, r);
+                        case MODULO -> type.modulo(l, r);
+                        case MULTIPLY -> type.multiply(l, r);
+                        case SUBTRACT -> type.subtract(l, r);
+
+                        case BIT_AND -> type.and(l, r);
+                        case BIT_OR -> type.or(l, r);
+                        case BIT_XOR -> type.xor(l, r);
+
+                        case BIT_SHIFT_LEFT -> type.lshift(l, r);
+                        case BIT_SHIFT_RIGHT -> type.rshift(l, r);
+                        case BIT_SHIFT_RIGHT_UNSIGNED -> type.rshiftUnsigned(l, r);
+                    };
+                }
+
+                if (lhs instanceof String lS && rhs instanceof String rS && binaryExpression.operator == BinaryExpression.Operator.ADD) {
+                    return lS + rS;
+                }
+
+                throw new IllegalArgumentException("Cannot resolve expression: " + binaryExpression + ". Operands of type " + lhs.getClass() + " and " + rhs.getClass() + " do not support operator " + binaryExpression.operator);
+            }
+
+            throw new IllegalArgumentException("Unknown Expression of type " + expression.getClass() + ": " + expression);
+        }
+
+        private static Object cast(Object in, DataType type) {
+            return switch (type) {
+                case BYTE -> ((Number) in).byteValue();
+                case CHAR -> Character.valueOf((char)((Number) in).byteValue());
```

Comment:
A `char` is 16 bits, not 8.
```suggestion
                case CHAR -> Character.valueOf((char)((Number) in).shortValue());
```